### ISSUE TITLE
Panic on 32-bit ARM Linux

### ISSUE THREAD

**Issue by petevine** [pijul/pijul#2]
One of the test panicks:

```rust
failures:

---- commands::test::with_changes_sth_to_record stdout ----
        thread 'commands::test::with_changes_sth_to_record' panicked at 'called `Result::unwrap()` on an `Err` value: Repository(IO(Error { repr: Os { code: 12, message: "Cannot allocate memory" } }))', src/libcore/result.rs:746
stack backtrace:
   1: 0xb6e94d7b - sys::backtrace::tracing::imp::write::h5fa72351789a20eb4nu
   2: 0xb6e98607 - panicking::default_handler::_$u7b$$u7b$closure$u7d$$u7d$::closure.42804
   3: 0xb6e98083 - panicking::default_handler::h5c31f6d656fe4f20yHy
   4: 0xb6e7fd7f - sys_common::unwind::begin_unwind_inner::h292ecb37a0508c3bCgt
   5: 0xb6e806af - sys_common::unwind::begin_unwind_fmt::h89eb0d66b9a05c98Ift
   6: 0xb6e94417 - rust_begin_unwind
   7: 0xb6ea322f - panicking::panic_fmt::ha2c4cab053dc37a4qYL
   8: 0xb6b98523 - result::unwrap_failed::h5036073332133961343
                at src/libcore/macros.rs:29
   9: 0xb6b99cbf - result::Result<T, E>::unwrap::h1460828402106608249
                at src/libcore/result.rs:687
  10: 0xb6b9a0a7 - commands::test::with_changes_sth_to_record::h59e86efc0ac53a11Krd
                at src/commands/test.rs:68
  11: 0xb6bd822f - boxed::F.FnBox<A>::call_box::h5562849258013620048
  12: 0xb6bda6e3 - sys_common::unwind::try::try_fn::h12826635515105541658
````

Compiled with `1.8` from yesterday.

It seems to be `ARM` specific as 32-bit `x86` Linux is fine.

**Comment by petevine** [comment]
Hey, it crashes on `1.5` too but it's puzzling to see different tests fail:

````
running 9 tests
test commands::escape::unix::test_escape ... ok
test commands::test::add_remove_nothing_to_record ... FAILED
test commands::test::init_creates_repo ... ok
test commands::test::in_empty_dir_nothing_to_record ... ok
test commands::test::init_nested_forbidden ... ok
test commands::test::no_remove_without_add ... ok
test commands::test::init_nested_allowed ... ok
gpg: gpg: no default secret key: secret key not available
no default secret key: secret key not available
gpg: signing failed: secret key not available
gpg: signing failed: secret key not available
test commands::test::with_changes_sth_to_record ... ok
test commands::test::add_record_pull ... ok
````

 and then:

````
running 9 tests
test commands::escape::unix::test_escape ... ok
test commands::test::init_creates_repo ... ok
test commands::test::add_remove_nothing_to_record ... ok
test commands::test::init_nested_forbidden ... ok
test commands::test::in_empty_dir_nothing_to_record ... ok
test commands::test::init_nested_allowed ... ok
test commands::test::no_remove_without_add ... ok
gpg: no default secret key: secret key not available
gpg: signing failed: secret key not available
test commands::test::add_record_pull ... ok
test commands::test::with_changes_sth_to_record ... FAILED
````

**Comment by pijul** [comment]
Seems related to the mmap in LMDB. Can you try tweaking the following line in libpijul/src/lib.rs:
```
try!(env.set_mapsize( (1 << 30) ));
```

(seems it is at line 207).
Set it to something smaller, such as 1<<15. What's the maximal value that works?

**Comment by petevine** [comment]
Going to `15` or below failed even more tests but `25` seems to work.

**Comment by petevine** [comment]
`29` works too - does it break 64-bit platforms?

**Comment by pijul** [comment]
Ok, thanks. What is the output of the following C program?
```C
#include<unistd.h>
#include<stdio.h>
int main(){ printf("%d\n",sysconf(_SC_PAGE_SIZE)); }
```

**Comment by petevine** [comment]
It's `4096`.

**Comment by pijul** [comment]
Alright, I'm actively working on it. There are other issues in our backend (LMDB), which forced us to rewrite one. We're getting there, but it's obviously taking much more time than initially planned.

**Comment by petevine** [comment]
Good to hear.  I'm going to give it a shot again once it's ready.

**Comment by petevine** [comment]
Any news?

**Comment by pijul** [comment]
Yes. We rewrote the database backend, into a crate called Sanakirja. It's pure Rust, although the code is a little bit ugly at the moment, for several reasons:

- We chose the alignment of things to be able to cast pointers directly, but it doesn't work that well with slices, which store their size as element count (as opposed to bytes).

- The prototype used lots of recursive functions, and so does the current version (modeled after the prototype), but now that dust has settled we might want to write loops over a stack-allocated stack instead.

- Its base layer is a transactional memory allocator, which kind of competes with the Rust allocator. I wonder what could be done to merge them into a single custom allocator.
